library/alloc/src/vec/mod.rs

```diff
@@ -1248,6 +1271,16 @@ impl<T, A: Allocator> Vec<T, A> {
     /// Modifying the vector may cause its buffer to be reallocated,
     /// which would also make any pointers to it invalid.
     ///
+    /// This method guarantees that when it is called multiple times without
+    /// the buffer being reallocated in the mean time, the returned pointer will
+    /// always be exactly the same, even for the purpose of the aliasing model, where
+    /// pointers may be invalidated even when the actual memory does not move.
+    /// Note that calling other methods that materialize references to the slice,
+    /// or references to specific elements you are planning on accessing through this pointer,
+    /// may still invalidate this pointer.
+    /// See the second example below for how this guarantee can be used.
```

Comment:
```suggestion
    /// This method guarantees that when it is called multiple times without
    /// the buffer being reallocated in the mean time, the returned pointers can
    /// all be used without interfering with each other -- there are no aliasing
    /// restrictions among these pointers. See the second example below for
    /// what this means in practice.
    /// However, calling other methods that materialize references to the slice
    /// or to particular elements may will still invalidate the pointer.
```

Reply:
One concern is that people could think they can now do this
```rust
fn main() { unsafe {
    let mut v = vec![0];
    let ptr1 = &mut *v.as_mut_ptr();
    let ptr2 = &mut *v.as_mut_ptr();
    *ptr1 = 0;
} }
```
because the returned pointers do not invalidate each other... or do you think it is sufficiently obvious that the source of invalidation is the `&mut`, not the `as_mut_ptr`?

Reply:
I think this is understandable paired with the stacked borrows model, but

It is probably good in the long run to have specific terminology that APIs like this can use.